src/compiler/checker.ts

```diff
@@ -9768,7 +9797,7 @@ namespace ts {
         }
 
         function getPropertiesOfType(type: Type): Symbol[] {
-            type = getApparentType(type);
+            type = getApparentType(getReducedType(type));

```

Comment:
Should `getReducedType` just be the first thing done in `getApparentType`? Feels like it should be pretty much always required.

Reply:
I tried that early on, but had to back it out. It had a noticeable performance impact on the compiler-with-unions and material-ui code bases, I suspect because we'd do a lot of reduction on contextual types that really isn't necessary (and in fact might be undesirable because we want contextual types to just have as much information as possible).

Reply:
Wait, does that mean that fake-nevers can introduce literal contexts and stuff right now? What's the error for
```ts
const z: {kind: number, field: "ok"} & {kind: string, member: "ok"} = {kind: 0 as never; field: "not ok", member: "not ok"};
```
?
So long as we're skipping the memberwise elaboration and just saying "{obj} is not assignable to never", it's ok, but if it's something like "string is not assignable to "ok"", that'd be not great.

Reply:
First off, we don't consider a property to be a discriminant unless one of the possible values is a literal type. But if I change your example to

```ts
const z: { kind: 1, field: "ok" } & { kind: 2, member: "ok" } = { kind: 0 as never, field: "not ok", member: "not ok" };
```

then we report `Type 'string' is not assignable to type 'never'` on each of `field` and `member`. I think that is a reasonable experience.

Reply:
We get errors on the individual properties above (instead of on the whole type) because we elaborate into types even when the target type is `never`. This is existing behavior, but it's probably a better experience not to elaborate on `never`.